molvis/molvispython/python/processors/MolecularStructureSource.py

```diff
@@ -0,0 +1,166 @@
+ # (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ # LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ # ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ # (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ # SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ #
+ #################################################################################
+
+import inviwopy as ivw
+
+from inviwopy.glm import *
+import ivwmolvis
+
+from Bio.PDB.MMCIFParser import MMCIFParser
+from Bio.PDB.PDBParser import PDBParser
+from Bio.PDB import CaPPBuilder
+
+import numpy as np
+
+
+class MolecularStructureSource(ivw.Processor):
+    def __init__(self, id, name):
+        ivw.Processor.__init__(self, id, name)
+        
+        self.addOutport(ivwmolvis.MolecularStructureOutport("molecule"))
+
+        self.filename = ivw.properties.FileProperty("filename", "Structure Filename", "", "molecularstructure")
+        self.filename.addNameFilter(ivw.properties.FileExtension("cif", "mmCIF file"))
+        self.filename.addNameFilter(ivw.properties.FileExtension("pdb", "PDB file"))
+        self.addProperty(self.filename, owner=False)
+
+        self.dataset = ivw.properties.StringProperty("dataset", "Data")
+        self.addProperty(self.dataset, owner=False)
+        self.dataset.readOnly = True
+
+    @staticmethod
+    def processorInfo():
+        return ivw.ProcessorInfo(
+            classIdentifier = "org.inviwo.MolecularStructureSource", 
+            displayName = "Molecular Structure Source",
+            category = "MolVis",
+            codeState = ivw.CodeState.Stable,
+            tags = ivw.Tags.PY
+        )
+
+    def getProcessorInfo(self):
+        return MolecularStructureSource.processorInfo()
+
+    def initializeResources(self):
+        pass
+
+    def process(self):
+        if not self.filename.value or not os.path.isfile(self.filename.value):
+            return
+
+        self.dataset.value = self.filename.value.split("/")[-1]
+
+        self.outports.molecule.setData(self.parseCIF(self.filename.value))
+
+    @staticmethod
+    def parseCIF(filename):
+        mesh = ivw.data.Mesh()
+
+        ext = filename.split(".")[-1]
+        if ext == 'gz' or ext == 'bz2':
+            ext = filename.split(".")[-2]
+
+        if (ext.startswith('pdb')):
+            parser = PDBParser(PERMISSIVE=1)
+        elif (ext.startswith('cif')):
+            # initialize mmcif parser
+            parser = MMCIFParser()
+        else:
+            ivw.LogError("unsupported extension '{}'".format(ext))
+            return mesh
+
+        structureName = filename.split("/")[-1]
+        structure = parser.get_structure(structureName, filename)
+
+        pos = []
+        bfactors = []
+        modelId = []
+        chainId = []
+        residueId = []
+        atomFullName = []
+
+        modelDict = dict()
```

Comment:
modelDict = {} also works

Reply:
I know. But for personal reference I find `dict()` a bit more meaningful.